src/Shared/test/MockTimeProvider.cs

```diff
@@ -0,0 +1,73 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Testing;
+
+public class MockTimeProvider : TimeProvider
+{
+    private readonly long _timestampFrequency =
+        // 10_000_000; // Windows
+        // 100_000_000; // 8 zeros, in between
+        // 1_000_000_000; // Linux
+        Stopwatch.Frequency;
+    private long _utcTicks;
+    private long _timestamp;
+
+    public MockTimeProvider()
+    {
+        // Use a random DateTimeOffset to ensure tests that incorrectly use the current DateTimeOffset fail always instead of only rarely.
+        var tenYears = TimeSpan.FromDays(365 * 10).Ticks;
+        _utcTicks = DateTimeOffset.UtcNow.Ticks + Random.Shared.NextInt64(-tenYears, tenYears);
+        // Timestamps often measure system uptime.
+        _timestamp = Random.Shared.NextInt64(0, System.GetTimestamp() * 100);
+    }
+
+    public MockTimeProvider(DateTimeOffset now)
+    {
+        _utcTicks = now.Ticks;
+        // Timestamps often measure system uptime.
+        _timestamp = Random.Shared.NextInt64(0, System.GetTimestamp() * 100);
+    }
+
+    public override DateTimeOffset GetUtcNow()
+    {
+        UtcNowCalled++;
+        return new DateTimeOffset(Interlocked.Read(ref _utcTicks), TimeSpan.Zero);
+    }
+
+    public override long GetTimestamp() => Interlocked.Read(ref _timestamp);
+
+    public override long TimestampFrequency => _timestampFrequency;
+
+    public int UtcNowCalled { get; private set; }
+
+    public void Advance(TimeSpan timeSpan)
+    {
+        Interlocked.Add(ref _utcTicks, timeSpan.Ticks);
+        Interlocked.Add(ref _timestamp, timeSpan.Ticks * (_timestampFrequency / TimeSpan.TicksPerSecond));
+    }
+
+    public void AdvanceTo(DateTimeOffset newUtcNow)
```

Comment:
Should the AdvanceTo calls check that the time doesn't go backwards?

Reply:
Backwards is fine, except if you make the timestamp go negative. We could check for that.

Reply:
Right, that's why I added a check for negative _timestamp and _utcTicks.

Reply:
That makes it worse. Because now a test will be flaky if it does any sort of time reversing.

And I still haven't seen a scenario where we do want to go backwards. So why let that happen until such a scenario occurs?

Reply:
Hmmm that's a good point about flakyness. I'll see if preventing reversals causes any issues.